### PR TITLE
CP-15244: Add framework for set IP address featrue

### PR DESCRIPTION
**Pull request by cheng-z**
This pull request is to:
1. Fix deadlock mentioned in https://lkml.org/lkml/2016/5/17/196
2. Fix watch event issue as if we watch on a dir, all change event of the dir keys need to go into the same chain.
3. Add Directory interface so we can query the subkeys
4. Add framework to set IP address 
5. Operations like read can return error as file not exist, add error return in DO logic

### CODE REVIEW

File: xenstoreclient/xenstore.go [hunk 1]
```diff
@@ -177,6 +192,11 @@ func (wq *WatchQueueManager) GetEventByKey(key string) (token string, ok bool) {
 func (wq *WatchQueueManager) AddChanByKey(key string) {
 	wq.rwlocker.Lock()
 	defer wq.rwlocker.Unlock()
+	for k := range wq.watchQueues {
```
Comment (phusl) [inline xenstoreclient/xenstore.go]: Maybe
```
if _, ok := wq.watchQueues[key]; ok {
//...
}
```

is more clear

File: xenstoreclient/xenstore.go [hunk 2]
```diff
@@ -534,7 +534,6 @@ func (xs *CachedXenStore) Clear() {
 
 func getDevPath() (devPath string, err error) {
 	devPaths := []string{
-		"/proc/xen/xenbus",
```
Comment (phusl) [inline xenstoreclient/xenstore.go]: Would better to mentio https://lkml.org/lkml/2016/5/17/196 in this commit message

File: xenstoreclient/xenstore.go (hunk 1 above)
Comment (cheng-z) [inline xenstoreclient/xenstore.go]: It is automatically cleaned by: gofmt -s -w .

File: xenstoreclient/xenstore.go (hunk 2 above)
Comment (cheng-z) [inline xenstoreclient/xenstore.go]: Thanks. I will add it.

**Comment by phusl** [comment]
Thanks. Merge to `IP-Setting` branch.

**Pull request merged**